### DISCUSSION TITLE
Add support for OpenGL/3D Acceleration (egl-headless, virgl) in QEMU to remotely view output through SPICE

### DISCUSSION THREAD

**Discussion by FGD-Garuda** [cockpit-project/cockpit-machines#547] [Ideas]
Page: Virtual Machines

--- Feature description ---

Hello!

This project has a LOT of potential. Your web interface/front-end is gorgeous too! With releases every 2 weeks that's quite agile and should help bring in new features all the time.

I'm hoping to contribute with the following feature request:

There is currently no convenient way in Cockpit to enable OpenGL/3D Acceleration to view the output through SPICE (as well as NoVNC but this one has limitations explained later below), both locally and remotely.

I've heard about using PCI Passthrough, if that works with Cockpit, but it has many limitations (like shutting down the screen when you are local) and opens up all kinds of other non-fixable issues for many use cases. Therefore 3D Acceleration through SPICE has become a well wanted solution for those use cases.

More and more OSes/Apps are leveraging OpenGL/3D Acceleration capabilities which makes this feature through SPICE high on demand, because SPICE is flexible, easy to use, doesn't require the guest to be connected to the network nor an RDP-like service running once the guest has booted up. Therefore SPICE gives access to GRUB menu, boot sequence and BIOS, just like a physical machine. SPICE also has unlimited window geometry (size) as it auto-resizes to your liking and needs and an excellent mouse integration. A simple Googling will return countless posts of people asking for this with other softwares. There's actually a bug in Virt-Manager 3.2 about that as 3D used to work before. I don't know if the bug has been logged, but here we're talking about Cockpit.

As for NoVNC it is tied into the browser front-end and therefore lacks the flexibility of SPICE, which is why I am focusing on SPICE here but the feature doesn't need to be dependent of it.

Support for that feature is already included in the Linux kernel since at least 4.16 I believe. Since a long time ago, Type 2 HyperVisors like VirtualBox already enable this feature by one single click (for both Linux and Windows guests), no other configuration required. As mentioned above Virt-Manager also has this feature, a bit more clicks, but it's there (buggy right now, that's a different story).

Also, to prove it works outside of Cockpit by using QEMU/KVM/libvirt, running locally the guest machine inside a Desktop Environment with the following command works perfectly:

`sudo qemu-system-x86_64 -name [domain-name] -drive file=/var/lib/libvirt/images/[domain-name].qcow2 -m 5G -enable-kvm -M q35 -cpu host -smp 6,sockets=1,cores=6,threads=1 -display sdl,gl=on -device virtio-vga,virgl=on -spice unix=on,disable-ticketing=on,image-compression=off,addr=[domain-name].sock,seamless-migration=on`

The key is the use of `-display sdl,gl=on` and `-device virtio-vga,virgl=on`. Other attributes may vary depending on your VM configuration, I am not sure about those following `-spice`.

Immediately a SPICE window, with auto-resize feature if supported by the guest, opens up and OpenGL with 3D Acceleration is working, making many modern Linux distros blistering fast and giving an unprecedented user experience.

Without this feature, those distros are laggy, some barely usable, and lack some compositor effects that require OpenGL/3D Acceleration. Even Windows 10/11 slows down a little bit. Some softwares, like rendering or games, require this too.

It is also possible to launch the VM with `egl-headless`display instead of `sdl` and then issue a remote-viewer command line to connect to the domain-name socket. This allows to pass 3D capabilities to the remote machine, although that is probably where most of the implementation challenge is. Working locally is one thing and probably much easier, working remotely is more challenging, but Cockpit is meant to work and maintain remotely.

Could you please consider including this feature in your backlog in order to help getting closer to achieve one of Cockpit's goal: making it intended to everyone, including those using OpenGL/3D acceleration.

**Comment by martinpitt** [comment]
Thanks for your idea! Just to set the expectations, this is not something which we'll work on soon. Cockpit is primarily a server administration tool, 3D support in VMs sounds more suitable in a desktop application like [GNOME Boxes](https://apps.gnome.org/app/org.gnome.Boxes/). You can't plumb through such a SPICE view in the browser anyway -- you'll need an external tool (such as spice viewer).

That said, if someone wants to integrate this into the project, we are certainly happy to provide some guidance and help.

**Comment by FGD-Garuda** [comment]
Yes the remote machine requires proper SPICE and virt-viewer packages to be installed. That works fine with Cockpit on the remote machine, except there is no 3D support of course.

If it was as simple as taking some courses to acquire the expertise I would probably do it, but I doubt it is that simple. lolll

Let's see if someone chimes in to help or let's wait for that day when Cockpit maintainers will have it on their priority list...

Thanks for the clarification! :)

**Comment by dkaschl** [comment]
I agree with you. The idea of quickly setting up virtual machines and supporting hardware acceleration would be a nice feature. I am actually working on this idea, but sadly haven't managed to get it running. Did you manage to have a headless linux host with a gpu, running linux guest vms utilizing virtio hard acceleration and connecting to the vm remotely via spice or RDP?

**Comment by FGD-Garuda** [comment]
> Did you manage to have a headless linux host with a gpu, running linux guest vms utilizing virtio hard acceleration and connecting to the vm remotely via spice or RDP?

Not that far. I do use Linux Host/Guests, virtio, headless, 3D acceleration and SPICE, but no GPU passthrough and totally outside Cockpit, I use it with Proxmox.

**Reply by dkaschl** [reply]
Thanks for your response! So this means you don't need or have a X server on your host machine, and you can actually access your host gpu in a headless setup? Would you mind sharing your _graphics_ and _video_ configuration of your guest? (=> virsh edit <guest name> should show this information.)

**Comment by FGD-Garuda** [comment]
> Thanks for your response! So this means you don't need or have a X server on your host machine, and you can actually access your host gpu in a headless setup?

Absolutely. I am running Host on Wayland and most Guests. I do have some X11 guests but most are Wayland.

> Would you mind sharing your _graphics_ and _video_ configuration of your guest? (=> virsh edit should show this information.)

From Proxmox?

**Reply by dkaschl** [reply]
Sorry my bad, I thought you are operating your host completely headless (no X, no Wayland) and you still got video hardware acceleration working. I am not sure how to check/see if video hardware acceleration is working on the guests.

Proxmox most likely will use the underlying libvirt module. The latter should offer the CLI for managing virtual machines called "virsh". But if Proxmox virtual machine description is in XML (https://libvirt.org/formatdomain.html#element-and-attribute-overview) then it should be the identical format.

**Reply by FGD-Garuda** [reply]
Well I badly responded.

The machine I am running SPICE from is under Wayland.
The machine that has Proxmox installed onto and HOSTS the VMs is totally headless.
The VMs running inside the above machine use mostly Wayland and some X (sorry Elon Musk I am not talking about you).

Not sure the info is in XML, I have to find the command I've seen in the past but it was not in XML, although maybe a different command returns same data in XML. I believe it was `qemu-server` command. I'll give a try this w-e as the Proxmox machine is not home this week.

**Reply by dkaschl** [reply]
Thanks for clearing it up. That would really be a headless host, fantastic!

Thanks a lot for your help! Appreciate it a lot!

**Reply by FGD-Garuda** [reply]
Ok so `virsh` does not seem to be the trick with Proxmox but anyway there are other ways to get the config, although it's pure text format.

One of my machine has this:

~~~
agent: 1
args: -display egl-headless,gl=core
balloon: 0
bios: seabios
boot: order=scsi0;ide2
cores: 6
cpu: host,flags=+ibpb;+virt-ssbd;+amd-ssbd;+pdpe1gb;+aes
ide2: none,media=cdrom
memory: 5192
meta: creation-qemu=6.1.1,ctime=1645583017
name: Garuda-wayland
net0: virtio=5A:C9:28:42:77:D0,bridge=vmbr0,firewall=1
numa: 0
ostype: l26
scsi0: local-lvm:vm-102-disk-0,cache=writeback,discard=on,size=40G,ssd=1
scsihw: virtio-scsi-single
smbios1: uuid=3009a862-4404-4aba-b471-2040984d3b98
sockets: 1
spice_enhancements: foldersharing=1,videostreaming=all
vga: virtio-gl,memory=512
vmgenid: 26eb2450-40f9-4e53-a4a3-584ed6c66af9
~~~

It's the `egl` and `virtio-gl` pieces that do the trick to enable better 3D acceleration. It ain't 100% powerful it's quite better than none, for example the wobbly windows Plasma effect works, which it does not at ALL when you don't have those options enabled.

I think the options have been improved since I opened the feature 1.5 years ago, but I have not had time to come back to it and see if I can make it even smoother. It's already very smooth to me considering they are VMs.

I can run them through SPICE or RDP.

**Reply by dkaschl** [reply]
Thanks a lot for the configuration information! It's a good guideline and although it doesn't like identical to the virsh config file, I think I can get the important configuration aspects out of this output too.
I assume it's not only 3d acceleration but also 2d? Like when browsing and scrolling the page there is no stuttering, or when watching videos on youtube, it's going pretty close up to 30fps?
Ideally I want to have a remote desktop that runs like a local one, of course I have a certain level of network throughput and latency, meaning LAN or WAN with very good latency and dl/ul speeds.

**Reply by FGD-Garuda** [reply]
Yes it's both 3D and 2D, usually I believe when 3D is enabled 2D is as well.

Videos on Youtube through SPICE you won't get 30fps, even in a small window. With the current config I have it's not a good experience to watch Tubes through SPICE. I haven't played with these configs for a long time and I know they have improved over time, maybe by tweaking them a little more you can get better, but I'm not sure you will get 30fps through SPICE.

It stutters a little bit when browsing pages, the heavier they are (ads, video snippets, animated GIFs, etc.) the worse it will get. But then again, I have not tweaked those settings.

If you want something very close to physical machine, you would have to use GPU Passthrough feature. On a headless host it's not a problem to dedicate the GPU to the VMs. However sometimes, depending on your hardware, it may be difficult to enable this feature. I have succeeded only once in 3 tries and it was on a 14yo laptop... But the other 2 machines were MiniPCs and maybe they are not meant for that.

Proxmox has a dedicated wiki page on how to turn on GPU Passthrough, I remember seeing it.

Still, how does GPU Passthrough works with SPICE, I have no idea, I was using RDP for that.

**Comment by javaarchive** [comment]
![image](https://github.com/cockpit-project/cockpit-machines/assets/20248577/ab2e6a8b-9e95-4286-8cd4-f14a946e1eed)
I translated the above proxmox into the xml for virtio, and it seems like if you have a egl headless "display" in the xml and at least a virtio display, a vnc device can see the 3D output I think. 
There was a minor glitch during boot though,
![image](https://github.com/cockpit-project/cockpit-machines/assets/20248577/6f4e124e-cb93-4209-ae99-29241c101e74)
So
```xml
<graphics type="egl-headless">
      <gl rendernode="/dev/dri/renderD128"/>
    </graphics>
```
and
```xml
<video>
      <model type="virtio" heads="1" primary="yes">
        <acceleration accel3d="yes"/>
      </model>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x01" function="0x0"/>
    </video>
```
These needed to be added to the xml manually. 

The vnc works on the web as well unsuprisingly
![image](https://github.com/cockpit-project/cockpit-machines/assets/20248577/b2072f13-5b24-46c1-9b9b-a5f56d813a93)
so now your experience is likely to more be dependent on vnc lag, than the capabilities of the guest graphics

**Comment by Janhouse** [comment]
Is the functionality to enable headless 3d acceleration missing in libvirt, libvirt-dbus or in this package?

virtio-gl is needed not just for 3d graphics, it is also needed for VA-API video transcoding hardware acceleration and that is needed in servers as well. Currently adjusting the xml manually it doesn't seem like the necessary qemu parameters are generated to enable gl.
